File: src/lib/cloud/analyze.js
// Architecture rules engine for Ground0: Cloud — Architecture Studio.
//
// Pure: analyzeArchitecture({ nodes, edges }) -> {
//   findings: [{ id, level, category, title, detail, nodeIds }],
//   securityScore, designScore, overall, verdict
// }
// Nodes are self-describing (they carry `kind`, `tier`, `ports`,
// `openToInternet`, `config`), so this module never touches the catalog.

import { STATEFUL_KINDS, COMPUTE_KINDS, PUBLIC_ENTRY_KINDS, SENSITIVE_PORTS } from '../../data/cloud/components.js'

const LEVEL_WEIGHT = { critical: 30, high: 17, warn: 9, info: 4, ok: 0 }
const SECURITY_CATS = ['security', 'iam', 'storage']

export const CATEGORY_LABELS = {
  security: 'Security',
  iam: 'IAM',
  storage: 'Storage',
  network: 'Network',
  reliability: 'Reliability',
  performance: 'Performance',
  cost: 'Cost',
}

const isStateful = (n) => STATEFUL_KINDS.includes(n.kind)
const isCompute = (n) => COMPUTE_KINDS.includes(n.kind)
const isInternet = (n) => n.kind === 'internet'

// Build undirected + directed adjacency once.
function buildGraph(nodes, edges) {
  const out = {} // directed: from -> [to]
  const inc = {} // directed: to -> [from]
  const undirected = {} // either direction
  nodes.forEach((n) => {
    out[n.id] = []
    inc[n.id] = []
    undirected[n.id] = []
  })
  edges.forEach((e) => {
    if (!out[e.from] || !out[e.to]) return
    out[e.from].push(e.to)
    inc[e.to].push(e.from)
    undirected[e.from].push(e.to)
    undirected[e.to].push(e.from)
  })
  return { out, inc, undirected }
}

// Set of node ids reachable from any internet node by following edges.
function reachableFromInternet(nodes, graph) {
  const entries = nodes.filter(isInternet).map((n) => n.id)
  const seen = new Set(entries)
  const stack = [...entries]
  while (stack.length) {
    const id = stack.pop()
    for (const next of graph.out[id] || []) {
      if (!seen.has(next)) {
        seen.add(next)
        stack.push(next)
      }
    }
  }
  return seen
}

export function analyzeArchitecture({ nodes = [], edges = [] }) {
  const findings = []
  const byId = Object.fromEntries(nodes.map((n) => [n.id, n]))
  const graph = buildGraph(nodes, edges)
  const reachable = reachableFromInternet(nodes, graph)

  const add = (level, category, title, detail, nodeIds = []) =>
    findings.push({ id: `${category}-${findings.length}`, level, category, title, detail, nodeIds })

  // ── SECURITY ────────────────────────────────────────────────────

  // 1. Stateful data in a non-private subnet.
  for (const n of nodes) {
    if (isStateful(n) && n.tier !== 'private') {
      add(
        'critical',
        'security',
        `${n.name} is not in a private subnet`,
        `Databases and caches should never sit in a "${n.tier}" tier. Move ${n.name} into a private subnet so it is unreachable from the internet.`,
        [n.id],
      )
    }
  }

  // 2. Stateful data directly reachable from the internet (1 hop) or open to 0.0.0.0/0.
  for (const n of nodes) {
    if (!isStateful(n)) continue
    const directFromEdge = graph.inc[n.id].some((src) => {
      const s = byId[src]
      return s && (isInternet(s) || PUBLIC_ENTRY_KINDS.includes(s.kind))
    })
    if (n.openToInternet || directFromEdge) {
      add(
        'critical',
        'security',
        `${n.name} is exposed to the internet`,
        n.openToInternet
          ? `${n.name}'s security group allows 0.0.0.0/0. Data tiers must only accept traffic from the application tier.`
          : `${n.name} receives traffic straight from an internet-facing component. Put a compute/app tier in front of it.`,
        [n.id],
      )
    }
  }

  // 3. Sensitive ports open to the whole internet.
  for (const n of nodes) {
    if (!n.openToInternet) continue
    const bad = (n.ports || []).filter((p) => SENSITIVE_PORTS[p])
    if (bad.length) {
      const names = bad.map((p) => `${SENSITIVE_PORTS[p]} (${p})`).join(', ')
      add(
        bad.some((p) => p !== 22 && p !== 3389) ? 'critical' : 'high',
        'security',
        `${n.name} opens ${names} to 0.0.0.0/0`,
        `Administrative and database ports must not be open to the whole internet. Restrict the security group to a bastion/VPN or the app tier.`,
        [n.id],
      )
    }
  }

  // 4. Compute exposed directly to the internet (no LB / gateway in front).
  for (const n of nodes) {
    if (!isCompute(n)) continue
    const directFromInternet = graph.inc[n.id].some((src) => byId[src] && isInternet(byId[src]))
    if (n.openToInternet || directFromInternet) {
      add(
        'high',
        'security',
        `${n.name} is directly internet-facing`,
        `Put a load balancer or API gateway in front of ${n.name} instead of exposing the instance. The compute tier should live in a private subnet.`,
        [n.id],
      )
    }
  }

  // 4b. Frontend/client talking straight to a data tier (no backend between).
  for (const n of nodes) {
    if (!isStateful(n)) continue
    const fromClient = graph.inc[n.id].some((src) => byId[src]?.kind === 'client')
    if (fromClient) {
      add(
        'high',
        'security',
        `${n.name} is queried directly by the frontend`,
        `A client / frontend is wired straight to ${n.name}. Put a backend / API tier in between — clients must never hold database credentials or hit the database directly.`,
        [n.id],
      )
    }
  }

  // 5. No WAF in front of a public entry point.
  const publicEntries = nodes.filter((n) => PUBLIC_ENTRY_KINDS.includes(n.kind) && reachable.has(n.id))
  const hasWaf = nodes.some((n) => n.kind === 'waf')
  if (publicEntries.length && !hasWaf) {
    add(
      'warn',
      'security',
      'No Web Application Firewall',
      'Public entry points are exposed without a WAF. Add one to filter common web exploits and abusive traffic.',
      publicEntries.map((n) => n.id),
    )
  }

  // 6. Plaintext-only edge at the front door (HTTP without HTTPS).
  for (const e of edges) {
    const from = byId[e.from]
    const to = byId[e.to]
    if (!from || !to) continue
    if ((isInternet(from) || from.kind === 'edge') && e.port === 80) {
      add(
        'warn',
        'security',
        'Traffic enters over plain HTTP',
        `The edge ${from.name} → ${to.name} uses port 80. Serve over HTTPS (443) and redirect HTTP so credentials are never sent in the clear.`,
        [from.id, to.id],
      )
    }
  }

  // 7. Over-privileged IAM role (wildcard *).
  for (const n of nodes) {
    if ((isCompute(n) || n.kind === 'serverless') && n.config?.iam === 'admin') {
      add(
        'high',
        'iam',
        `${n.name} uses an over-privileged IAM role`,
        `${n.name} is attached to an admin / wildcard (*:*) policy. Grant least privilege — scope the role to only the actions and resources it actually needs.`,
        [n.id],
      )
    }
  }

  // 8. Encryption at rest.
  for (const n of nodes) {
    if (n.config && 'encrypted' in n.config && !n.config.encrypted) {
      add(
        n.kind === 'database' ? 'high' : 'warn',
        'storage',
        `${n.name} is not encrypted at rest`,
        `Turn on encryption at rest (KMS) for ${n.name}. It is a one-click setting and is expected for anything storing data.`,
        [n.id],
      )
    }
  }

  // 9. Public object storage.
  for (const n of nodes) {
    if (n.kind === 'storage' && n.config?.public) {
      add(
        'critical',
        'storage',
        `${n.name} is a public bucket`,
        `${n.name} allows public access. Enable "Block Public Access" and serve files via CloudFront with an origin access identity instead.`,
        [n.id],
      )
    }
  }

  // 10. Prefer SSM Session Manager over open SSH.
  const sshNodes = nodes.filter((n) => (n.ports || []).includes(22))
  if (sshNodes.length) {
    add(
      'info',
      'security',
      'Prefer SSM Session Manager over SSH',
      'A component exposes port 22 (SSH). In production, use AWS Systems Manager Session Manager for shell access — no inbound port to open, and every session is audit-logged.',
      sshNodes.map((n) => n.id),
    )
  }

  // ── RELIABILITY / PERFORMANCE / NETWORK / COST ──────────────────

  const computeNodes = nodes.filter(isCompute)

  // 7. Multiple compute nodes but nothing balancing them.
  const hasLb = nodes.some((n) => n.kind === 'lb' || n.kind === 'gateway')
  if (computeNodes.length >= 2 && !hasLb) {
    add(
      'warn',
      'reliability',
      'No load balancer for multiple compute targets',
      'You have more than one compute component but no load balancer. Add one so traffic is distributed and a single instance failing does not take the app down.',
      computeNodes.map((n) => n.id),
    )
  }

  // 8. Single-instance compute = single point of failure.
  for (const n of computeNodes) {
    const instances = n.config?.instances ?? (n.kind === 'serverless' ? 99 : 1)
    if (!n.config?.autoScale && instances <= 1) {
      add(
        'warn',
        'reliability',
        `${n.name} is a single point of failure`,
        `${n.name} runs a single instance with no auto scaling. Use an Auto Scaling Group across multiple AZs for resilience and to handle load spikes.`,
        [n.id],
      )
    }
  }

  // 9. Database without Multi-AZ.
  for (const n of nodes) {
    if (n.kind === 'database' && n.config && 'multiAz' in n.config && !n.config.multiAz) {
      add(
        'info',
        'reliability',
        `${n.name} is not Multi-AZ`,
        `Enable Multi-AZ on ${n.name} so a standby in another availability zone can take over automatically if the primary fails.`,
        [n.id],
      )
    }
  }

  // Automated backups on relational databases.
  for (const n of nodes) {
    if (n.type === 'rds' && n.config && 'backups' in n.config && !n.config.backups) {
      add(
        'warn',
        'reliability',
        `${n.name} has no automated backups`,
        `Enable automated snapshots / point-in-time recovery on ${n.name}, or a failure means permanent data loss.`,
        [n.id],
      )
    }
  }

  // Private compute with no NAT Gateway for outbound traffic.
  const privateCompute = computeNodes.filter((n) => n.tier === 'private')
  const hasNat = nodes.some((n) => n.kind === 'nat')
  if (privateCompute.length && !hasNat) {
    add(
      'warn',
      'network',
      'No NAT Gateway for private subnets',
      'Private instances cannot reach the internet for OS patches, package installs or third-party APIs without a NAT Gateway (or VPC endpoints) for outbound traffic.',
      privateCompute.map((n) => n.id),
    )
  }

  // Burstable dev-size instances under sustained production traffic.
  for (const n of computeNodes) {
    if (n.config?.instanceType === 't3.micro' && !n.config?.autoScale && reachable.has(n.id)) {
      add(
        'info',
        'performance',
        `${n.name} runs on a burstable t3.micro`,
        `t3.micro is a small burstable type — fine for dev, risky for sustained production load. Pick a larger type or enable Auto Scaling, then run a load test to size it.`,
        [n.id],
      )
    }
  }

  // Read path with no cache / CDN.
  const hasDatabase = nodes.some((n) => n.kind === 'database')
  const hasCache = nodes.some((n) => n.kind === 'cache')
  const hasCdn = nodes.some((n) => n.kind === 'cdn')
  if (hasDatabase && reachable.size > 1 && !hasCache && !hasCdn) {
    add(
      'info',
      'performance',
      'No caching layer',
      'Add a CDN for static content and/or an in-memory cache (ElastiCache) in front of the database to cut latency and read load.',
      [],
    )
  }

  // 11. Orphan nodes (placed but wired to nothing).
  for (const n of nodes) {
    if (isInternet(n)) continue
    if ((graph.undirected[n.id] || []).length === 0) {
      add(
        'warn',
        'reliability',
        `${n.name} is not connected`,
        `${n.name} has no connections, so no traffic can reach it. Wire it into the flow or remove it.`,
        [n.id],
      )
    }
  }

  // 12. Services present but no entry point at all.
  if (nodes.length > 0 && !nodes.some(isInternet) && (hasLb || computeNodes.length)) {
    add(
      'info',
      'reliability',
      'No internet entry point',
      'Drop a "Users / Internet" node and connect it to your front door so you can simulate a real request end-to-end.',
      [],
    )
  }

  // ── SCORING ─────────────────────────────────────────────────────
  // Security score penalises security-family findings (incl. IAM + storage);
  // design score penalises everything else (reliability / perf / network / cost).
  const penalty = (keep) =>
    findings.filter(keep).reduce((s, f) => s + (LEVEL_WEIGHT[f.level] || 0), 0)

  const isSecurityCat = (c) => SECURITY_CATS.includes(c)
  const securityScore = clamp(100 - penalty((f) => isSecurityCat(f.category)))
  const designScore = clamp(100 - penalty((f) => !isSecurityCat(f.category)))
  const overall = Math.round((securityScore + designScore) / 2)

  let verdict
  const hasCritical = findings.some((f) => f.level === 'critical')
  if (hasCritical) verdict = { label: 'Insecure', tone: 'bad' }
  else if (overall >= 90) verdict = { label: 'Well-architected', tone: 'good' }
  else if (overall >= 70) verdict = { label: 'Solid, with gaps', tone: 'warn' }
  else verdict = { label: 'Needs work', tone: 'warn' }

  return { findings, securityScore, designScore, overall, verdict }
}

function clamp(v) {
  return Math.max(0, Math.min(100, Math.round(v)))
}

export const FINDING_STYLES = {
  critical: { text: 'text-rose-300', dot: 'bg-rose-500', badge: 'bg-rose-500/15 text-rose-300', label: 'Critical' },
  high: { text: 'text-orange-300', dot: 'bg-orange-500', badge: 'bg-orange-500/15 text-orange-300', label: 'High' },
  warn: { text: 'text-amber-300', dot: 'bg-amber-500', badge: 'bg-amber-500/15 text-amber-300', label: 'Warning' },
  info: { text: 'text-sky-300', dot: 'bg-sky-500', badge: 'bg-sky-500/15 text-sky-300', label: 'Tip' },
  ok: { text: 'text-emerald-300', dot: 'bg-emerald-500', badge: 'bg-emerald-500/15 text-emerald-300', label: 'OK' },
}

File: go-server/internal/studio/cloud/analyze.go
package cloud

import (
	"fmt"

	"ground0.dev/goserver/internal/model"
)

var levelWeight = map[string]int{"critical": 30, "high": 17, "warn": 9, "info": 4, "ok": 0}
var securityCats = map[string]bool{"security": true, "iam": true, "storage": true}
var correctnessCats = map[string]bool{"correctness": true}

type Finding struct {
	ID       string   `json:"id"`
	Level    string   `json:"level"`
	Category string   `json:"category"`
	Title    string   `json:"title"`
	Detail   string   `json:"detail"`
	NodeIDs  []string `json:"nodeIds"`
}

type Verdict struct {
	Label string `json:"label"`
	Tone  string `json:"tone"`
}

type AnalyzeResult struct {
	Findings         []Finding `json:"findings"`
	CorrectnessScore int       `json:"correctnessScore"`
	SecurityScore    int       `json:"securityScore"`
	DesignScore      int       `json:"designScore"`
	Overall          int       `json:"overall"`
	Verdict          Verdict   `json:"verdict"`
	Empty            bool      `json:"empty"`
}

type adjacency struct {
	out, inc, undirected map[string][]string
}

func buildGraph(nodes []model.Node, edges []model.Edge) adjacency {
	a := adjacency{out: map[string][]string{}, inc: map[string][]string{}, undirected: map[string][]string{}}
	for _, n := range nodes {
		a.out[n.ID] = []string{}
		a.inc[n.ID] = []string{}
		a.undirected[n.ID] = []string{}
	}
	for _, e := range edges {
		if _, ok := a.out[e.From]; !ok {
			continue
		}
		if _, ok := a.out[e.To]; !ok {
			continue
		}
		a.out[e.From] = append(a.out[e.From], e.To)
		a.inc[e.To] = append(a.inc[e.To], e.From)
		a.undirected[e.From] = append(a.undirected[e.From], e.To)
		a.undirected[e.To] = append(a.undirected[e.To], e.From)
	}
	return a
}

func reachableFromInternet(nodes []model.Node, g adjacency) map[string]bool {
	seen := map[string]bool{}
	var stack []string
	for _, n := range nodes {
		if isInternetNode(n) {
			seen[n.ID] = true
			stack = append(stack, n.ID)
		}
	}
	for len(stack) > 0 {
		id := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		for _, next := range g.out[id] {
			if !seen[next] {
				seen[next] = true
				stack = append(stack, next)
			}
		}
	}
	return seen
}

// findingAdder accumulates findings in encounter order, matching analyze.js's
// sequential `add()` calls (finding ids embed a running per-category index).
type findingAdder struct {
	findings []Finding
}

func (a *findingAdder) add(level, category, title, detail string, nodeIDs []string) {
	if nodeIDs == nil {
		nodeIDs = []string{}
	}
	a.findings = append(a.findings, Finding{
		ID: fmt.Sprintf("%s-%d", category, len(a.findings)), Level: level, Category: category,
		Title: title, Detail: detail, NodeIDs: nodeIDs,
	})
}

func configBool(cfg map[string]any, key string) (bool, bool) {
	if cfg == nil {
		return false, false
	}
	v, ok := cfg[key]
	if !ok {
		return false, false
	}
	b, _ := v.(bool)
	return b, true
}

func configString(cfg map[string]any, key string) string {
	if cfg == nil {
		return ""
	}
	s, _ := cfg[key].(string)
	return s
}

func configNumber(cfg map[string]any, key string) (float64, bool) {
	if cfg == nil {
		return 0, false
	}
	switch v := cfg[key].(type) {
	case float64:
		return v, true
	case int:
		return float64(v), true
	}
	return 0, false
}

// AnalyzeArchitecture ports src/lib/cloud/analyze.js's analyzeArchitecture
// rule-by-rule. See that file for the rationale behind each check.
func AnalyzeArchitecture(g model.Graph) AnalyzeResult {
	nodes, edges := g.Nodes, g.Edges

	if len(nodes) == 0 {
		return AnalyzeResult{
			Findings: []Finding{}, Verdict: Verdict{Label: "Empty canvas", Tone: "warn"}, Empty: true,
		}
	}

	byID := map[string]model.Node{}
	for _, n := range nodes {
		byID[n.ID] = n
	}
	graph := buildGraph(nodes, edges)
	reachable := reachableFromInternet(nodes, graph)
	fa := &findingAdder{findings: []Finding{}}

	// ── TOPOLOGY: illogical / impossible wiring ──
	invalidCount := 0
	for _, e := range edges {
		from, ok1 := byID[e.From]
		to, ok2 := byID[e.To]
		if !ok1 || !ok2 {
			continue
		}
		res := classifyEdge(from, to)
		if res.Level == "illegal" {
			invalidCount++
			fa.add("critical", "correctness", fmt.Sprintf("Illogical connection: %s → %s", from.DisplayName(), to.DisplayName()),
				res.Reason, []string{from.ID, to.ID})
		}
	}

	// ── SECURITY ──

	// 1. Stateful data in a non-private subnet.
	for _, n := range nodes {
		if isStateful(n) && n.Tier != "private" {
			fa.add("critical", "security", n.DisplayName()+" is not in a private subnet",
				fmt.Sprintf("Databases and caches should never sit in a %q tier. Move %s into a private subnet so it is unreachable from the internet.", n.Tier, n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 2. Stateful data directly reachable from the internet (1 hop) or open to 0.0.0.0/0.
	for _, n := range nodes {
		if !isStateful(n) {
			continue
		}
		directFromEdge := false
		for _, src := range graph.inc[n.ID] {
			s, ok := byID[src]
			if ok && (isInternetNode(s) || publicEntryKinds[s.Kind]) {
				directFromEdge = true
				break
			}
		}
		if n.OpenToInternet || directFromEdge {
			detail := fmt.Sprintf("%s receives traffic straight from an internet-facing component. Put a compute/app tier in front of it.", n.DisplayName())
			if n.OpenToInternet {
				detail = fmt.Sprintf("%s's security group allows 0.0.0.0/0. Data tiers must only accept traffic from the application tier.", n.DisplayName())
			}
			fa.add("critical", "security", n.DisplayName()+" is exposed to the internet", detail, []string{n.ID})
		}
	}

	// 3. Sensitive ports open to the whole internet.
	for _, n := range nodes {
		if !n.OpenToInternet {
			continue
		}
		var bad []int
		for _, p := range n.Ports {
			if _, ok := sensitivePorts[p]; ok {
				bad = append(bad, p)
			}
		}
		if len(bad) > 0 {
			names := ""
			nonSSHOrRDP := false
			for i, p := range bad {
				if i > 0 {
					names += ", "
				}
				names += fmt.Sprintf("%s (%d)", sensitivePorts[p], p)
				if p != 22 && p != 3389 {
					nonSSHOrRDP = true
				}
			}
			level := "high"
			if nonSSHOrRDP {
				level = "critical"
			}
			fa.add(level, "security", fmt.Sprintf("%s opens %s to 0.0.0.0/0", n.DisplayName(), names),
				"Administrative and database ports must not be open to the whole internet. Restrict the security group to a bastion/VPN or the app tier.",
				[]string{n.ID})
		}
	}

	// 4. Compute exposed directly to the internet (no LB / gateway in front).
	for _, n := range nodes {
		if !isCompute(n) {
			continue
		}
		directFromInternet := false
		for _, src := range graph.inc[n.ID] {
			if s, ok := byID[src]; ok && isInternetNode(s) {
				directFromInternet = true
				break
			}
		}
		if n.OpenToInternet || directFromInternet {
			fa.add("high", "security", n.DisplayName()+" is directly internet-facing",
				fmt.Sprintf("Put a load balancer or API gateway in front of %s instead of exposing the instance. The compute tier should live in a private subnet.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 4b. Frontend/client talking straight to a data tier.
	for _, n := range nodes {
		if !isStateful(n) {
			continue
		}
		fromClient := false
		for _, src := range graph.inc[n.ID] {
			if s, ok := byID[src]; ok && s.Kind == "client" {
				fromClient = true
				break
			}
		}
		if fromClient {
			fa.add("high", "security", n.DisplayName()+" is queried directly by the frontend",
				fmt.Sprintf("A client / frontend is wired straight to %s. Put a backend / API tier in between — clients must never hold database credentials or hit the database directly.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 5. No WAF in front of a public entry point.
	var publicEntries []model.Node
	for _, n := range nodes {
		if publicEntryKinds[n.Kind] && reachable[n.ID] {
			publicEntries = append(publicEntries, n)
		}
	}
	hasWaf := false
	for _, n := range nodes {
		if n.Kind == "waf" {
			hasWaf = true
			break
		}
	}
	if len(publicEntries) > 0 && !hasWaf {
		ids := make([]string, len(publicEntries))
		for i, n := range publicEntries {
			ids[i] = n.ID
		}
		fa.add("warn", "security", "No Web Application Firewall",
			"Public entry points are exposed without a WAF. Add one to filter common web exploits and abusive traffic.", ids)
	}

	// 6. Plaintext-only edge at the front door (HTTP without HTTPS).
	for _, e := range edges {
		from, ok1 := byID[e.From]
		to, ok2 := byID[e.To]
		if !ok1 || !ok2 {
			continue
		}
		if (isInternetNode(from) || from.Kind == "edge") && e.Port == 80 {
			fa.add("warn", "security", "Traffic enters over plain HTTP",
				fmt.Sprintf("The edge %s → %s uses port 80. Serve over HTTPS (443) and redirect HTTP so credentials are never sent in the clear.", from.DisplayName(), to.DisplayName()),
				[]string{from.ID, to.ID})
		}
	}

	// 7. Over-privileged IAM role (wildcard *).
	for _, n := range nodes {
		if (isCompute(n) || n.Kind == "serverless") && configString(n.Config, "iam") == "admin" {
			fa.add("high", "iam", n.DisplayName()+" uses an over-privileged IAM role",
				fmt.Sprintf("%s is attached to an admin / wildcard (*:*) policy. Grant least privilege — scope the role to only the actions and resources it actually needs.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 8. Encryption at rest.
	for _, n := range nodes {
		if enc, present := configBool(n.Config, "encrypted"); present && !enc {
			level := "warn"
			if n.Kind == "database" {
				level = "high"
			}
			fa.add(level, "storage", n.DisplayName()+" is not encrypted at rest",
				fmt.Sprintf("Turn on encryption at rest (KMS) for %s. It is a one-click setting and is expected for anything storing data.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 9. Public object storage.
	for _, n := range nodes {
		if pub, _ := configBool(n.Config, "public"); n.Kind == "storage" && pub {
			fa.add("critical", "storage", n.DisplayName()+" is a public bucket",
				fmt.Sprintf("%s allows public access. Enable \"Block Public Access\" and serve files via CloudFront with an origin access identity instead.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 10. Prefer SSM Session Manager over open SSH.
	var sshNodes []string
	for _, n := range nodes {
		for _, p := range n.Ports {
			if p == 22 {
				sshNodes = append(sshNodes, n.ID)
				break
			}
		}
	}
	if len(sshNodes) > 0 {
		fa.add("info", "security", "Prefer SSM Session Manager over SSH",
			"A component exposes port 22 (SSH). In production, use AWS Systems Manager Session Manager for shell access — no inbound port to open, and every session is audit-logged.",
			sshNodes)
	}

	// ── RELIABILITY / PERFORMANCE / NETWORK / COST ──

	var computeNodes []model.Node
	for _, n := range nodes {
		if isCompute(n) {
			computeNodes = append(computeNodes, n)
		}
	}

	hasLb := false
	for _, n := range nodes {
		if n.Kind == "lb" || n.Kind == "gateway" {
			hasLb = true
			break
		}
	}
	if len(computeNodes) >= 2 && !hasLb {
		ids := make([]string, len(computeNodes))
		for i, n := range computeNodes {
			ids[i] = n.ID
		}
		fa.add("warn", "reliability", "No load balancer for multiple compute targets",
			"You have more than one compute component but no load balancer. Add one so traffic is distributed and a single instance failing does not take the app down.", ids)
	}

	// 8. Single-instance compute = single point of failure.
	for _, n := range computeNodes {
		instances, hasInstances := configNumber(n.Config, "instances")
		if !hasInstances {
			instances = 1
			if n.Kind == "serverless" {
				instances = 99
			}
		}
		autoScale, _ := configBool(n.Config, "autoScale")
		if !autoScale && instances <= 1 {
			fa.add("warn", "reliability", n.DisplayName()+" is a single point of failure",
				fmt.Sprintf("%s runs a single instance with no auto scaling. Use an Auto Scaling Group across multiple AZs for resilience and to handle load spikes.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 9. Database without Multi-AZ.
	for _, n := range nodes {
		if n.Kind != "database" {
			continue
		}
		if multiAz, present := configBool(n.Config, "multiAz"); present && !multiAz {
			fa.add("info", "reliability", n.DisplayName()+" is not Multi-AZ",
				fmt.Sprintf("Enable Multi-AZ on %s so a standby in another availability zone can take over automatically if the primary fails.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// Automated backups on relational databases.
	for _, n := range nodes {
		if n.Type != "rds" {
			continue
		}
		if backups, present := configBool(n.Config, "backups"); present && !backups {
			fa.add("warn", "reliability", n.DisplayName()+" has no automated backups",
				fmt.Sprintf("Enable automated snapshots / point-in-time recovery on %s, or a failure means permanent data loss.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// Private compute with no NAT Gateway for outbound traffic.
	var privateCompute []model.Node
	for _, n := range computeNodes {
		if n.Tier == "private" {
			privateCompute = append(privateCompute, n)
		}
	}
	hasNat := false
	for _, n := range nodes {
		if n.Kind == "nat" {
			hasNat = true
			break
		}
	}
	if len(privateCompute) > 0 && !hasNat {
		ids := make([]string, len(privateCompute))
		for i, n := range privateCompute {
			ids[i] = n.ID
		}
		fa.add("warn", "network", "No NAT Gateway for private subnets",
			"Private instances cannot reach the internet for OS patches, package installs or third-party APIs without a NAT Gateway (or VPC endpoints) for outbound traffic.", ids)
	}

	// Burstable dev-size instances under sustained production traffic.
	for _, n := range computeNodes {
		autoScale, _ := configBool(n.Config, "autoScale")
		if configString(n.Config, "instanceType") == "t3.micro" && !autoScale && reachable[n.ID] {
			fa.add("info", "performance", n.DisplayName()+" runs on a burstable t3.micro",
				"t3.micro is a small burstable type — fine for dev, risky for sustained production load. Pick a larger type or enable Auto Scaling, then run a load test to size it.",
				[]string{n.ID})
		}
	}

	hasDatabase := false
	hasCache := false
	hasCdn := false
	var cacheIDs []string
	for _, n := range nodes {
		switch n.Kind {
		case "database":
			hasDatabase = true
		case "cache":
			hasCache = true
			cacheIDs = append(cacheIDs, n.ID)
		case "cdn":
			hasCdn = true
		}
	}

	// A cache is never the system of record — it must have a database behind it.
	if hasCache && !hasDatabase {
		if cacheIDs == nil {
			cacheIDs = []string{}
		}
		fa.add("high", "correctness", "Cache has no database behind it",
			"A cache (Redis / ElastiCache) is volatile and only speeds up reads — it cannot be the source of truth. Add a database for the cache to read through (cache-aside) and to persist data; a cache can never stand alone as the data tier.",
			cacheIDs)
	}

	if hasDatabase && len(reachable) > 1 && !hasCache && !hasCdn {
		fa.add("info", "performance", "No caching layer",
			"Add a CDN for static content and/or an in-memory cache (ElastiCache) in front of the database to cut latency and read load.", nil)
	}

	// 11. Orphan nodes (placed but wired to nothing).
	for _, n := range nodes {
		if isInternetNode(n) {
			continue
		}
		if len(graph.undirected[n.ID]) == 0 {
			fa.add("warn", "correctness", n.DisplayName()+" is not connected",
				fmt.Sprintf("%s has no connections, so no traffic can reach it. Wire it into the flow or remove it.", n.DisplayName()),
				[]string{n.ID})
		}
	}

	// 12. Services present but no entry point at all.
	hasInternetEntry := false
	for _, n := range nodes {
		if isInternetNode(n) {
			hasInternetEntry = true
			break
		}
	}
	if len(nodes) > 0 && !hasInternetEntry && (hasLb || len(computeNodes) > 0) {
		fa.add("info", "correctness", "No internet entry point",
			"Drop a \"Users / Internet\" node and connect it to your front door so you can simulate a real request end-to-end.", nil)
	}

	// ── SCORING ──
	penalty := func(keep func(Finding) bool) int {
		s := 0
		for _, f := range fa.findings {
			if keep(f) {
				s += levelWeight[f.Level]
			}
		}
		return s
	}
	correctnessScore := clampScore(100 - penalty(func(f Finding) bool { return correctnessCats[f.Category] }))
	securityScore := clampScore(100 - penalty(func(f Finding) bool { return securityCats[f.Category] }))
	designScore := clampScore(100 - penalty(func(f Finding) bool {
		return !securityCats[f.Category] && !correctnessCats[f.Category]
	}))
	overall := int(roundHalfAwayFromZero(float64(correctnessScore+securityScore+designScore) / 3))

	hasCritical := false
	for _, f := range fa.findings {
		if f.Level == "critical" {
			hasCritical = true
			break
		}
	}
	var verdict Verdict
	switch {
	case invalidCount > 0:
		verdict = Verdict{Label: "Invalid setup", Tone: "bad"}
	case hasCritical:
		verdict = Verdict{Label: "Insecure", Tone: "bad"}
	case overall >= 90:
		verdict = Verdict{Label: "Well-architected", Tone: "good"}
	case overall >= 70:
		verdict = Verdict{Label: "Solid, with gaps", Tone: "warn"}
	default:
		verdict = Verdict{Label: "Needs work", Tone: "warn"}
	}

	return AnalyzeResult{
		Findings: fa.findings, CorrectnessScore: correctnessScore, SecurityScore: securityScore,
		DesignScore: designScore, Overall: overall, Verdict: verdict, Empty: false,
	}
}

func clampScore(v int) int {
	if v < 0 {
		return 0
	}
	if v > 100 {
		return 100
	}
	return v
}

func roundHalfAwayFromZero(v float64) float64 {
	if v < 0 {
		return -roundHalfAwayFromZero(-v)
	}
	return float64(int(v + 0.5))
}
